src/mam4xx/mo_sethet.hpp

```diff
@@ -176,6 +183,303 @@ void find_ktop(
+    //                             path below the cloud
+    //                        dt = dz(cm)/um(cm/s)
+    //-----------------------------------------------------------------
+    team.team_barrier();
+    xdtm = delz(kk) / xum; // the traveling time in each dz
+
+    xxx2 = (xh2o2(kk) - xgas2(kk));
+    if (xxx2 != 0.0) { // if no washout lifetime = 1.e29
+      yh2o2 = xh2o2(kk) / xxx2 * xdtm;
+    } else {
+      yh2o2 = large_value_lifetime;
+    }
+    tmp_hetrates[1](kk) =
+        haero::max(1.0 / yh2o2, 0.0) * stay; // FIXME: bad constant index
+
+    xxx3 = (xso2(kk) - xgas3(kk));
+    if (xxx3 != 0.0) { // if no washout lifetime = 1.e29
+      yso2 = xso2(kk) / xxx3 * xdtm;
+    } else {
+      yso2 = large_value_lifetime;
+    }
+    tmp_hetrates[2](kk) =
+        haero::max(1.0 / yso2, 0.0) * stay; // FIXME: bad constant index
+  }
+
+  //-----------------------------------------------------------------
+  //       ... part 2, in-cloud solve for low henry constant
+  //                   hno3 and h2o2 have both in and under cloud
+  //-----------------------------------------------------------------
+  team.team_barrier();
+  constexpr int local_gas_pcnst = gas_pcnst;
+  for (int kk = ktop; kk < pver; kk++) {
+    bool skip = false;
+    for (int mm = 0; mm < local_gas_pcnst; ++mm) {
+
+      if (rain(kk) <= 0.0) {
+        het_rates[mm](kk) = 0.0;
+        skip = true;
+      }
+    }
+    if (skip)
+      continue;
+
+    work1 = avo2 * xliq(kk);
+    work2 = const0 * tfld(kk);
+
+    if (h2o2_ndx >= 0) {
+      calc_het_rates(satf_h2o2, rain(kk), xhen_h2o2(kk), // in
+                     tmp_hetrates[1](kk), work1, work2,  // in
+                     het_rates[h2o2_ndx](kk));           // out
+    }
+
+    // if ( prog_modal_aero .and.
+    if (so2_ndx >= 0 && h2o2_ndx >= 0) {
+      het_rates[so2_ndx](kk) = het_rates[h2o2_ndx](kk);
+    } else if (so2_ndx >= 0) {
+      calc_het_rates(satf_so2, rain(kk), xhen_so2(kk),  // in
+                     tmp_hetrates[2](kk), work1, work2, // in
+                     het_rates[so2_ndx](kk));           // out
+    }
+
+    if (h2so4_ndx >= 0) {
+      calc_het_rates(satf_hno3, rain(kk), xhen_hno3(kk), // in
+                     tmp_hetrates[0](kk), work1, work2,  // in
+                     het_rates[h2so4_ndx](kk));          // out
+    }
+  }
+
+  //-----------------------------------------------------------------
+  //	... Set rates above tropopause = 0.
+  //-----------------------------------------------------------------
+  team.team_barrier();
+  bool abort = false;
+  for (int mm = 0; mm < gas_wetdep_cnt; mm++) {
+    int mm2 = wetdep_map[mm];
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, ktop),
+                         [&](int kk) { het_rates[mm2](kk) = 0.0; });
+    Kokkos::parallel_for(Kokkos::TeamVectorRange(team, local_pver),
+                         [&](int kk) {
+                           if (het_rates[mm2](kk) == MISSING)
```

Comment:
Is this just a sanity check?

Reply:
It looks like `abort` is being set on device and then checked on host. Some values are automatically copied host to device but never back to host.  A parallel_reduction is probably needed instead of a parallel_for.

Reply:
Gotcha, so I could do a parallel_reduce where if a `MISSING` value is found, add something to the cumulative variable, then on host if the var isn't 0 then abort?

Reply:
Yes, copy the logic that reduces total_rain and change to an int.